### PR TITLE
Async ui

### PR DESCRIPTION
**Pull request by MohamedShatry**
Add asynchronous functionality to the UI to show spinners while waiting for data

### CODE REVIEW

File: src/main/ngapp/src/app/main/main.component.html [hunk 1]
```diff
@@ -2,27 +2,27 @@
...
   </nb-layout-header>
 
   <ng-template #dialog let-data let-ref="dialogRef">
-    <nb-card class="dialog-template">
-      <nb-card-header class="header-container" (click)="ref.close()">
+    <nb-card class="dialog-template" [nbSpinner]="languageUpdateLoading" nbSpinnerStatus="primary">
+      <nb-card-header class="header-container">
         Update Language Preference
-        <button nbSuffix nbButton ghost autofocus>
+        <button nbSuffix nbButton ghost autofocus (click)="(ref.close() && closeAlert())">
```
Comment (kiprot) [inline src/main/ngapp/src/app/main/main.component.html]: will this close the pop up card automatically (one a user has successfully selected a language to change to)? 
i had noticed that the pop-up window stayed until a user clicked the 'x' close button even after language had been successfully changed.

File: src/main/ngapp/src/app/main/main.component.ts [hunk 2]
```diff
@@ -208,23 +239,33 @@ export class MainComponent implements OnInit {
       const messagePromise = this.fetchMessages(chatroomId);
       messagePromise
         .then((messages) => {
+          this.chatSectionLoading = true;
           console.log(messages);
         })
         .catch((err) => {
+          this.chatSectionLoading = true;
```
Comment (kiprot) [inline src/main/ngapp/src/app/main/main.component.ts:246]: why would the chatRoomSection be loading when there is an error?

File: src/main/ngapp/src/app/chat-data.service.ts [hunk 3]
```diff
@@ -16,16 +16,17 @@ export class ChatDataService {
...
-    return this.http.get<User[]>('/user?userId=' + localUser.uid);
+    return this.http.get<User>('/user?userId=' + localUser.uid);
   }
 
   // Create a new user object for the user.
   addUser(user: User): Observable<any> {
     return this.http.post<User>('/user', user);
   }
 
+  //
```
Comment (kiprot) [inline src/main/ngapp/src/app/chat-data.service.ts:29]: comment marks added but not comments made on line 21 :)

File: src/main/ngapp/src/app/main/main.component.html (hunk 1 above)
Comment (MohamedShatry) [reply src/main/ngapp/src/app/main/main.component.html]: The window stays open until a user closes it with the x buttton. After the language has been updated, we just show the user the updated message

File: src/main/ngapp/src/app/main/main.component.ts (hunk 2 above)
Comment (MohamedShatry) [reply src/main/ngapp/src/app/main/main.component.ts]: Should be false. Updated

File: src/main/ngapp/src/app/chat-data.service.ts (hunk 3 above)
Comment (MohamedShatry) [reply src/main/ngapp/src/app/chat-data.service.ts]: Comments added

File: src/main/java/com/google/sps/servlets/UserServlet.java [hunk 4]
```diff
@@ -117,8 +117,9 @@ public void doPut(HttpServletRequest request, HttpServletResponse response) thro
         String jsonString = IOUtils.toString((request.getInputStream()));
         User userInput = new Gson().fromJson(jsonString, User.class);
         String userID = userInput.userId;
-
+
```
Comment (octopus1000) [inline src/main/java/com/google/sps/servlets/UserServlet.java:120]: Please remove unnecessary empty indent/empty lines.

File: src/main/ngapp/src/app/chat-data.service.ts [hunk 5]
```diff
@@ -80,9 +81,7 @@ export class ChatDataService {
 
   // Add message in a new chatroom.
   addMessage(input: Post) {
-
-    const promise = this.http.post<Post>('/messages', input, this.httpOptions).toPromise();
-    promise.catch((err) => console.error('Error sending new message: ', err));
+    return this.http.post<Post>('/messages', input, this.httpOptions).toPromise();
```
Comment (octopus1000) [inline src/main/ngapp/src/app/chat-data.service.ts:84]: Specify return type
addMessage(input:Post): (some type)

File: src/main/ngapp/src/app/register/register.component.html [hunk 6]
```diff
@@ -102,7 +16,7 @@ <h4>Registration</h4>
         <nb-select placeholder="Select Showcase" ngModel #language="ngModel" name="language" id="language" required [status]="language.dirty? (language.invalid? 'danger': 'success'): 'basic'">
             <nb-option value="en">English</nb-option>
```
Comment (octopus1000) [inline src/main/ngapp/src/app/register/register.component.html:17]: Seeing this list showing up in two files. Consider move the language into a file and generate options dynaimcally.

File: src/main/ngapp/src/app/main/main.component.ts [hunk 7]
```diff
@@ -119,21 +127,42 @@ export class MainComponent implements OnInit {
...
+            console.log(res);
+        })
+        .catch((err) => {
+            this.languageUpdateLoading = false;
+            this.languageUpdateError = err;
+            this.languageUpdateFailure = true;
+            console.error(err);
+        });
+    }).catch(err => {
+        this.languageUpdateLoading = false;
```
Comment (octopus1000) [inline src/main/ngapp/src/app/main/main.component.ts:155]: Do you need both success and error?

!this.LanguageupdateLoading && !this.Failure seems to indicate success.

**Pull request closed without merging**